### DISCUSSION TITLE
using BlueSeer transformation in a REST context

### DISCUSSION THREAD

**Discussion by matthewlongfield** [blueseerERP/blueseer#141] [Q&A] [answered]
First off let me say that I am super appreciative for the work it takes, and the selflessness it shows to put that hard work out there for others to use and learn from, no strings attached. Hats off, I really admire your ethic and generosity!!!

I am most interested in using BlueSeer in "standalone translator" mode and in as much of a REST context, or scripting context calling the jars/classes as directly as possible.

For a first example, I'm attempting to call a translation upon an arbitrary input file, using a specific map, and writing the output file to specific location. My 1st attempt is to use EDIbs like so, from the powershell commandline on windows11, current directory C:\BlueSeer where I've installed BS from the binary installer:
.\jre17\bin\java -cp ".;c:/blueseer/dist/*" com.blueseer.edi.EDIbs -if C:\BlueSeer\edi\sampledata\ACME_850.txt -of C:\BlueSeer\edi\test\acmeorder.xml -m C:\BlueSeer\edi\maps\bs850ToXML.jar

This does kickoff the translation, picking up the sampledata\ACME_850.txt and creating a XML, BUT it doesn't respect the -of argument - it creates the files in edi\batch and edi\out using the edpd Attributes page FilePrefix+counter+FileSuffix entries

First question: How can I specify an overridden output filename in this context?

Second question: Is there a way to call the classes already available, in a REST context, supplying the X12 input from a POST method and returning the transformed XML in the response body, or using a *.bat file to call java -cp getting the input X12 text from stdin and sending the transformed text to stdout?

As I gain confidence and understanding of BlueSeer, I will happily share any code I create.

Thanks again for the fine work
MattL

**Comment by vaughnte** [comment, marked as answer]
Hey Matt,

First of all...thank you for the compliments.   Much appreciated.     You are correct about the EDIbs class issue ignoring the -of flag.   It's a bug.   The command line EDIbs class was done prior to some recent updates in config lookup and I failed to bake the changes back into EDIbs...as it is not used a lot.    I'll be uploading a new jar file immediately after posting this.   Download the blueseer.jar file from the following zip (https://github.com/blueseerERP/blueseer/releases/download/v7.0/blueseer.jaronly.zip) and overlay into your c:\blueseer\dist folder...overwriting the old....then try your command line again.    I also need to update the documentation as the allow parameters to EDIbs have changed slightly.   For the '-m' parameter....use only the name of the map...(without .jar) and do not use the full path.   Also, add a '-x' parameter to the end of the command string...as this is now required to distinguish cmd line processing in the engine.   Here's a working example:     `.\jre17\bin\java -cp ".;c:/blueseer/dist/*" com.blueseer.edi.EDIbs -if C:\BlueSeer\edi\sampledata\ACME_850.txt -of C:\BlueSeer\edi\test\acmeorder.xml.txt -m bs850ToXML -x`
Also....the -of path will now honor the /dirparentpath/filename value...ONLY if the /dirparentpath/ is a valid path.   Note...you can actually run it with any 'valid' -m map even though you may not have that specific record defined in edi_mstr.    I was able to run the above command line with -m bs850ToXML, -m bs850ToJSON, and -m bs850to856 ....changing the -of accordingly.....all against the same source -if file.    
To answer your other question....I don't have a complete web service to do an HTTP Post with the -if data and return the -of in a HTTPResponse at the moment....but all the pieces are there to make that happen.   I do have some coding in the works now with utilities.apiServer that will eventually allow this approach.    A lot of refactoring of the engine is needed for performance though...as there are way too many db reads from transaction start to finish to allow a reasonable HTTP response return time.

**Reply by matthewlongfield** [reply]
That's awesome, Terry. I can verify that the correction works and EDIbs performs as you've said after replacing those 2 jars in the blueseer dist folder. Unexpected goodies that the different maps (tried bs850ToJSON too) also produce their respective output. This is a VERY usable utility as it stands, but I'm drooling over the apiServer prospect too.
The only concerning thing I saw was a windows defender alert squawking about a Trojan:Java/Multiverze and quarantining bsmf.jar, but I'm quite sure it's safe so I just added the entire C:\BlueSeer folder as a scanning exception.  

I hope to digest/grok more of your code - past few days I've been grazing and I have a vague grasp of it, but I expect that EDI.processFile was probably not getting the outfile arg populated like it needed. Was the the config lookup the culprit? I do wish I had a better grasp of the complete flow and what classes get called in a particular command chain of events.
I can't ask you to waste your valuable time and effort spoon feeding me though, so I'll continue OR, and share any quirks I observe.

Please let me know if I can help with testing or such, especially as it relates to EDI X12 850/855/856/810 distribution type stuff - I've been working in home-improvement/hardware distribution for quite a while now but currently have time & effort to invest in opensource

Thanks again
MattL
